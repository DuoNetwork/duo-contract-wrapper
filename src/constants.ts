export const ETH_MAINNET_ID = 1;
export const ETH_KOVAN_ID = 42;

export const DUMMY_ADDR = '0x0';

export const CTD_INCEPTION = 'Inception';
export const CTD_TRADING = 'Trading';
export const CTD_PRERESET = 'PreReset';
export const CTD_RESET = 'Reset';
export const CTD_LOADING = 'Loading';

export const BTV_UP_RESET = 'UpwardReset';
export const BTV_DOWN_RESET = 'DownwardReset';
export const BTV_PERIOD_RESET = 'PeriodicReset';

export const ESP_NOT_STARTED = 'NotStarted';
export const ESP_MODERATOR = 'Moderator';
export const ESP_CONTRACT = 'Contract';

// wallet and contract
export const INCEPTION_BLK_KOVAN = 9355871;
export const INCEPTION_BLK_MAIN = 0;

// managed
export const EVENT_UPDATE_ROLE_MANAGER = 'UpdateRoleManager';
export const EVENT_UPDATE_OPERATOR = 'UpdateOperator';

// esplanade
export const EVENT_ADD_ADDRESS = 'AddAddress';
export const EVENT_REMOVE_ADDRESS = 'RemoveAddress';
export const EVENT_PROVIDE_ADDRESS = 'ProvideAddress';
export const EVENT_ADD_CUSTODIAN = 'AddCustodian';
export const EVENT_ADD_OTHER_CONTRACT = 'AddOtherContract';
export const EVENT_START_CONTRACT_VOTING = 'StartContractVoting';
export const EVENT_TERMINATE_CONTRACT_VOTING = 'TerminateContractVoting';
export const EVENT_START_MODERATOR_VOTING = 'StartModeratorVoting';
export const EVENT_TERMINATEBY_TIMEOUT = 'TerminateByTimeOut';
export const EVENT_VOTE = 'Vote';
export const EVENT_COMPLETE_VOTING = 'CompleteVoting';
export const EVENT_REPLACE_MODERATOR = 'ReplaceModerator';

// custodian
export const EVENT_START_TRADING = 'StartTrading';
export const EVENT_START_PRE_RESET = 'StartPreReset';
export const EVENT_START_RESET = 'StartReset';
export const EVENT_MATURED = 'Matured'
export const EVENT_ACCEPT_PRICE = 'AcceptPrice';
export const EVENT_CREATE = 'Create';
export const EVENT_REDEEM = 'Redeem';
export const EVENT_TOTAL_SUPPLY = 'TotalSupply';
export const EVENT_TRANSFER = 'Transfer';
export const EVENT_APPROVAL = 'Approval';
export const EVENT_COLLECT_FEE = 'CollectFee';
export const EVENT_UPDATE_ORACLE = 'UpdateOracle';
export const EVENT_UPDATE_FEE_COLLECTOR = 'UpdateFeeCollector';

// beethovan
export const EVENT_SET_VALUE = 'SetValue';

// magi
export const EVENT_COMMIT_PRICE = 'CommitPrice';
export const EVENT_UPDATE_PRICE_FEED = 'UpdatePriceFeed';

export const STATE_INCEPTION = '0';
export const STATE_TRADING = '1';
export const STATE_PRERESET = '2';
export const STATE_RESET = '3';
export const RESET_STATE_UP = '0';
export const RESET_STATE_DOWN = '1';
export const RESET_STATE_PERIOD = '2';
export const VOTING_NOT_STARTED = '0';
export const VOTING_MODERATOR = '1';
export const VOTING_CONTRACT = '2';
export const FN_START_CONTRACT = 'startContract';
export const FN_COMMIT_PRICE = 'commitPrice';
export const SRC_MYETHER = 'myether';
export const SRC_INFURA = 'infura';
export const PROVIDER_LOCAL_HTTP = 'http://localhost:8545';
// export const PROVIDER_LOCAL_WS = 'ws://localhost:8546';
export const PROVIDER_INFURA_MAIN_WS = 'wss://mainnet.infura.io/ws';
export const PROVIDER_INFURA_KOVAN_WS = 'wss://kovan.infura.io/ws';
export const PROVIDER_MYETHER_MAIN = 'https://api.myetherapi.com/eth';
export const PROVIDER_MYETHER_ROP = 'https://api.myetherapi.com/rop';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const TRANSFER_TOKEN_INTERVAL = 30; // in seconds
export const TRANSFER_TOKEN_GAS = 120000;
export const TRANSFER_TOKEN_GAS_TH = 0.01; // in ether
export const SET_VALUE_GAS = 50000;
export const SET_VALUE_GAS_PRICE = 3; // in Gwei
export const COLLECT_FEE_GAS = 40000;
export const COLLECT_FEE_GAS_PRICE = 4; // in Gwei
export const ADD_ADDR_GAS = 40000;
export const ADD_ADDR_GAS_PRICE = 5; // in Gwei
export const REMOVE_ADDR_GAS = 40000;
export const REMOVE_ADDR_GAS_PRICE = 5; // in Gwei
export const UPDATE_ADDR_GAS = 40000;
export const UPDATE_ADDR_GAS_PRICE = 5; // in Gwei
export const TRANSFER_INTERVAL = 10; // in seconds
export const TRANSFER_GAS_TH = 0.01;
export const REDEEM_INTERVAL = 10; // in seconds
export const REDEEM_GAS = 80000;
export const REDEEM_GAS_TH = 0.005;
export const CREATE_INTERVAL = 10; // in seconds
export const CREATE_GAS = 160000;
export const CREATE_GAS_TH = 0.01;
export const DEFAULT_GAS_PRICE = 5;
export const PRE_RESET_GAS_LIMIT = 120000;
export const RESET_GAS_LIMIT = 4000000;
export const EVENT_FETCH_BLOCK_INTERVAL = 100;
export const EVENT_FETCH_TIME_INVERVAL = 600000;

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const BTV_STATE = {
	LAST_OPERATION_TIME: 0,
	OPERATION_COOLDOWN: 1,
	STATE: 2,
	MIN_BALANCE: 3,
	TOTAL_SUPPLYA: 4,
	TOTAL_SUPPLYB: 5,
	ETH_COLLATERAL_INWEI: 6,
	NAVA_INWEI: 7,
	NAVB_INWEI: 8,
	LAST_PRICE_INWEI: 9,
	LAST_PRICETIME_INSECOND: 10,
	RESET_PRICE_INWEI: 11,
	RESET_PRICETIME_INSECOND: 12,
	CREATE_COMMINBP: 13,
	REDEEM_COMMINBP: 14,
	PERIOD: 15,
	MATURITY: 16,
	PRERESET_WAITING_BLOCKS: 17,
	PRICE_FETCH_COOLDOWN: 18,
	NEXT_RESET_ADDR_INDEX: 19,
	TOTAL_USERS: 20,
	FEE_BALANCE_INWEI: 21,
	RESET_STATE: 22,
	ALPHA_INBP: 23,
	BETA_INWEI: 24,
	PERIOD_COUPON_INWEI: 25,
	LIMIT_PERIODIC_INWEI: 26,
	LIMIT_UPPER_INWEI: 27,
	LIMIT_LOWER_INWEI: 28,
	ITERATION_GAS_THRESHOLD: 29
};
