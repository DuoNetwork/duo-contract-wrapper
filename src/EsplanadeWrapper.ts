import BaseContractWrapper from './BaseContractWrapper';
import * as CST from './constants';
import esplanadeAbi from './static/Esplanade.json';
import { IVotingData } from './types';
import util from './util';
import Web3Wrapper from './Web3Wrapper';

export default class EsplanadeWapper extends BaseContractWrapper {
	public readonly events = [
		CST.EVENT_ADD_ADDRESS,
		CST.EVENT_REMOVE_ADDRESS,
		CST.EVENT_PROVIDE_ADDRESS,
		CST.EVENT_ADD_CUSTODIAN,
		CST.EVENT_ADD_OTHER_CONTRACT,
		CST.EVENT_START_CONTRACT_VOTING,
		CST.EVENT_TERMINATE_CONTRACT_VOTING,
		CST.EVENT_START_MODERATOR_VOTING,
		CST.EVENT_TERMINATEBY_TIMEOUT,
		CST.EVENT_VOTE,
		CST.EVENT_COMPLETE_VOTING,
		CST.EVENT_REPLACE_MODERATOR
	];
	constructor(web3Wrapper: Web3Wrapper, address: string) {
		super(web3Wrapper, esplanadeAbi.abi, address);
	}

	public getAddressPoolIndex(hot: boolean) {
		return hot ? 1 : 0;
	}

	public static convertVotingStage(stage: string) {
		switch (stage.valueOf()) {
			case CST.VOTING_MODERATOR:
				return CST.ESP_MODERATOR;
			case CST.VOTING_CONTRACT:
				return CST.ESP_CONTRACT;
			default:
				return CST.ESP_NOT_STARTED;
		}
	}

	public async getVotingStage() {
		const stage = await this.contract.methods.votingStage().call();
		return EsplanadeWapper.convertVotingStage(stage.valueOf());
	}

	public getModerator(): Promise<string> {
		return this.contract.methods.moderator().call();
	}

	public getCandidate(): Promise<string> {
		return this.contract.methods.candidate().call();
	}

	public async getPoolAddresses(hot: boolean): Promise<string[]> {
		const poolIndex = this.getAddressPoolIndex(hot);
		const poolSizes = await this.contract.methods.getAddressPoolSizes().call();
		const length = Number(poolSizes[poolIndex].valueOf());
		const addresses: string[] = [];
		for (let i = 0; i < length; i++)
			addresses.push(await this.contract.methods.addrPool(poolIndex, i).call());

		return addresses;
	}

	public async getContractAddresses(custodian: boolean): Promise<string[]> {
		const poolSizes = await this.contract.methods.getContractPoolSizes().call();
		const length = Number(poolSizes[custodian ? 0 : 1].valueOf());
		const addresses: string[] = [];
		for (let i = 0; i < length; i++)
			addresses.push(
				custodian
					? await this.contract.methods.custodianPool(i).call()
					: await this.contract.methods.otherContractPool(i).call()
			);

		return addresses;
	}

	public async getOperationCoolDown() {
		return Number((await this.contract.methods.operatorCoolDown().call()).valueOf()) * 1000;
	}

	public async getLastOperationTime() {
		return Number((await this.contract.methods.lastOperationTime().call()).valueOf()) * 1000;
	}

	public async isStarted(): Promise<boolean> {
		return await this.contract.methods.started().call();
	}

	public async getVotingData(): Promise<IVotingData> {
		return {
			started:
				Number((await this.contract.methods.voteStartTimestamp().call()).valueof()) * 1000,
			votedFor: Number((await this.contract.methods.votedFor().call()).valueof()),
			votedAgainst: Number((await this.contract.methods.votedAgainst().call()).valueof()),
			totalVoters: Number(
				(await this.contract.methods.getAddressPoolSizes().call())[
					this.getAddressPoolIndex(false)
				].valueof()
			)
		};
	}

	public startContractVoting(address: string, candidateAddress: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.startContractVoting(candidateAddress).send({
			from: address
		});
	}

	public terminateContractVoting(address: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.terminateContractVoting().send({
			from: address
		});
	}

	public terminateByTimeout(address: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.terminateByTimeout().send({
			from: address
		});
	}

	public startModeratorVoting(address: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.startModeratorVoting().send({
			from: address
		});
	}

	public vote(address: string, voteFor: boolean) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.vote(voteFor).send({
			from: address
		});
	}

	public startManager(address: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.startManager().send({
			from: address
		});
	}

	public async startManagerRaw(
		address: string,
		privateKey: string,
		gasPrice: number,
		gasLimit: number,
		nonce: number = -1
	) {
		util.logInfo(`the account ${address} is starting esplanade`);

		const abi = {
			name: 'startManager',
			type: 'function',
			inputs: []
		};

		const command = this.web3Wrapper.generateTxString(abi, []);
		await this.sendTransactionRaw(
			address,
			privateKey,
			this.address,
			0,
			gasPrice,
			gasLimit,
			nonce,
			command
		);
	}

	public addCustodian(address: string, custodianAddr: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.addCustodian(custodianAddr).send({
			from: address
		});
	}

	public async addCustodianRaw(
		address: string,
		privateKey: string,
		custodianAddr: string,
		gasPrice: number,
		gasLimit: number,
		nonce: number = -1
	) {
		util.logInfo(`the account ${address} is addCustodian ${custodianAddr}`);
		const abi = {
			type: 'function',
			inputs: [
				{
					name: 'custodianAddr',
					type: 'address'
				}
			],
			name: 'addCustodian'
		};

		const input = [custodianAddr];
		const command = this.web3Wrapper.generateTxString(abi, input);
		await this.sendTransactionRaw(
			address,
			privateKey,
			this.address,
			0,
			gasPrice,
			gasLimit,
			nonce,
			command
		);
	}

	public addOtherContracts(address: string, contractAddr: string) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();

		return this.contract.methods.addOtherContracts(contractAddr).send({
			from: address
		});
	}

	public addAddress(address: string, addr1: string, addr2: string, hot: boolean) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();
		return this.contract.methods.addAddress(addr1, addr2, this.getAddressPoolIndex(hot)).send({
			from: address
		});
	}

	public removeAddress(address: string, addr: string, hot: boolean) {
		if (this.web3Wrapper.isReadOnly()) return this.web3Wrapper.readOnlyReject();
		return this.contract.methods.removeAddress(addr, this.getAddressPoolIndex(hot)).send({
			from: address
		});
	}
}
