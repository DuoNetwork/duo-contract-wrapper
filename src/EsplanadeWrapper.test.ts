import * as CST from './constants';
import EsplanadeWrapper from './EsplanadeWrapper';

const addrPool = [['coldAddr1', 'coldAddr2', 'coldAddr3'], ['hotAddr1', 'hotAddr2', 'hotAddr3']];
const custodianPool = ['custodianAddr1', 'custodianAddr2', 'custodianAddr3'];
const otherPool = ['otherAddr1', 'otherAddr2', 'otherAddr3'];

const web3Wrapper = {
	isReadOnly: jest.fn(() => true),
	onSwitchToMetaMask: jest.fn(() => ({} as any)),
	fromWei: jest.fn(value => value * 1e-18),
	onSwitchToLedger: jest.fn(() => ({} as any)),
	readOnlyReject: jest.fn(() => Promise.reject('Read Only Mode')),
	getEthBalance: jest.fn(() => Promise.resolve('10')),
	createContract: jest.fn(() => ({
		methods: {
			moderator: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('moderator'))
			})),
			candidate: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('candidatre'))
			})),
			votingStage: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('1'))
			})),
			addrPool: jest.fn((poolIndex, i) => ({
				call: jest.fn(() => Promise.resolve(addrPool[poolIndex][i]))
			})),
			getAddressPoolSizes: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve(['3', '3']))
			})),
			custodianPool: jest.fn(i => ({
				call: jest.fn(() => Promise.resolve(custodianPool[i]))
			})),
			otherContractPool: jest.fn(i => ({
				call: jest.fn(() => Promise.resolve(otherPool[i]))
			})),
			getContractPoolSizes: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve(['3', '3']))
			})),
			operatorCoolDown: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('100'))
			})),
			lastOperationTime: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('100'))
			})),
			started: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve(true))
			})),
			voteStartTimestamp: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve(123456789))
			})),
			votedFor: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('10'))
			})),
			votedAgainst: jest.fn(() => ({
				call: jest.fn(() => Promise.resolve('5'))
			})),
			startContractVoting: jest.fn(() => ({
				send: jest.fn()
			})),
			terminateContractVoting: jest.fn(() => ({
				send: jest.fn()
			})),
			terminateByTimeout: jest.fn(() => ({
				send: jest.fn()
			})),
			startModeratorVoting: jest.fn(() => ({
				send: jest.fn()
			})),
			vote: jest.fn(() => ({
				send: jest.fn()
			})),
			startManager: jest.fn(() => ({
				send: jest.fn()
			})),
			addCustodian: jest.fn(() => ({
				send: jest.fn()
			})),
			addOtherContracts: jest.fn(() => ({
				send: jest.fn()
			})),
			addAddress: jest.fn(() => ({
				send: jest.fn()
			})),
			removeAddress: jest.fn(() => ({
				send: jest.fn()
			})),
			getAddresses: jest.fn(() => ({
				call: jest.fn()
			}))
		}
	}))
} as any;
const esplanadeWrapper = new EsplanadeWrapper(web3Wrapper, 'address');

test('startContractVoting', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.startContractVoting('account', 'candidateAddr');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.startContractVoting('account', 'candidateAdddr');
	expect(
		(esplanadeWrapper.contract.methods.startContractVoting as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('terminateContractVoting', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.terminateContractVoting('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.terminateContractVoting('account');
	expect(esplanadeWrapper.contract.methods.terminateContractVoting as jest.Mock).toBeCalledTimes(
		1
	);
});

test('terminateByTimeout', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.terminateByTimeout('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.terminateByTimeout('account');
	expect(esplanadeWrapper.contract.methods.terminateByTimeout as jest.Mock).toBeCalledTimes(1);
});

test('startModeratorVoting', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.startModeratorVoting('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.startModeratorVoting('account');
	expect(esplanadeWrapper.contract.methods.startModeratorVoting as jest.Mock).toBeCalledTimes(1);
});

test('vote', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.vote('account', true);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.vote('account', true);
	await esplanadeWrapper.vote('', true);
	expect((esplanadeWrapper.contract.methods.vote as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startManager', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.startManager('account');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.startManager('account');
	expect(esplanadeWrapper.contract.methods.startManager as jest.Mock).toBeCalledTimes(1);
});

test('addCustodian', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.addCustodian('account', 'custodianAddr');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.addCustodian('account', 'custodianAddr');
	expect(
		(esplanadeWrapper.contract.methods.addCustodian as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('addOtherContracts', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.addOtherContracts('account', 'contractAddr');
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.addOtherContracts('account', 'otherContractAddr');
	expect(
		(esplanadeWrapper.contract.methods.addOtherContracts as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('addAddress', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.addAddress('account', 'addr1', 'addr2', true);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.addAddress('account', 'addr1', 'addr2', true);
	expect(
		(esplanadeWrapper.contract.methods.addAddress as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('removeAddress', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	try {
		await esplanadeWrapper.removeAddress('account', 'addr', true);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
	web3Wrapper.isReadOnly = jest.fn(() => false);
	await esplanadeWrapper.removeAddress('account', 'addr', true);
	expect(
		(esplanadeWrapper.contract.methods.removeAddress as jest.Mock).mock.calls
	).toMatchSnapshot();
});

test('getStates', async () => {
	web3Wrapper.isReadOnly = jest.fn(() => true);
	expect(await esplanadeWrapper.getStates()).toMatchSnapshot();
});

test('convertVotingStage', () => {
	expect(EsplanadeWrapper.convertVotingStage(CST.VOTING_CONTRACT)).toBe(CST.ESP_CONTRACT);
	expect(EsplanadeWrapper.convertVotingStage(CST.VOTING_MODERATOR)).toBe(CST.ESP_MODERATOR);
	expect(EsplanadeWrapper.convertVotingStage('any')).toBe(CST.ESP_NOT_STARTED);
});

test('getAddressPoolindex', () => {
	expect(esplanadeWrapper.getAddressPoolIndex(true)).toBe(1);
	expect(esplanadeWrapper.getAddressPoolIndex(false)).toBe(0);
});

test('getModerator', async () => {
	expect(await esplanadeWrapper.getModerator()).toMatchSnapshot();
});

test('getCandidate', async () => {
	expect(await esplanadeWrapper.getCandidate()).toMatchSnapshot();
});

test('getAddressPoolAddress', async () => {
	expect(await esplanadeWrapper.getAddressPoolAddress(true, 0)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolAddress(true, 1)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolAddress(true, 2)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolAddress(false, 0)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolAddress(false, 1)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolAddress(false, 2)).toMatchSnapshot();
});

test('getContractPoolAddress', async () => {
	expect(await esplanadeWrapper.getContractPoolAddress(true, 0)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolAddress(true, 1)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolAddress(true, 2)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolAddress(false, 0)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolAddress(false, 1)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolAddress(false, 2)).toMatchSnapshot();
});

test('getAddressPoolSize', async () => {
	expect(await esplanadeWrapper.getAddressPoolSize(true)).toMatchSnapshot();
	expect(await esplanadeWrapper.getAddressPoolSize(false)).toMatchSnapshot();
});

test('getContractPoolSize', async () => {
	expect(await esplanadeWrapper.getContractPoolSize(true)).toMatchSnapshot();
	expect(await esplanadeWrapper.getContractPoolSize(false)).toMatchSnapshot();
});

test('voteStartTimestamp', async () => {
	expect(await esplanadeWrapper.getVotingData()).toMatchSnapshot();
});
