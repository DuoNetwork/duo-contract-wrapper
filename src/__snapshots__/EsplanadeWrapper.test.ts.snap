// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAddress 1`] = `
Array [
  Array [
    "addr1",
    "addr2",
    1,
  ],
]
`;

exports[`addAddress, readOnly 1`] = `"Read Only Mode"`;

exports[`addCustodian 1`] = `
Array [
  Array [
    "custodianAddr",
  ],
]
`;

exports[`addCustodian, readOnly 1`] = `"Read Only Mode"`;

exports[`addOtherContracts 1`] = `
Array [
  Array [
    "otherContractAddr",
  ],
]
`;

exports[`addOtherContracts, readOnly 1`] = `"Read Only Mode"`;

exports[`getAddrs 1`] = `
Object {
  "candidate": Object {
    "address": "candidatre",
    "balance": "10",
  },
  "moderator": Object {
    "address": "moderator",
    "balance": "10",
  },
}
`;

exports[`getAddrs 2`] = `
Array [
  Array [
    "moderator",
  ],
  Array [
    "candidatre",
  ],
]
`;

exports[`getContractAddresses 1`] = `
Array [
  Object {
    "address": "otherAddr1",
    "balance": "10",
  },
  Object {
    "address": "otherAddr2",
    "balance": "10",
  },
  Object {
    "address": "otherAddr3",
    "balance": "10",
  },
]
`;

exports[`getContractAddresses 2`] = `
Array [
  Object {
    "address": "custodianAddr1",
    "balance": "10",
  },
  Object {
    "address": "custodianAddr2",
    "balance": "10",
  },
  Object {
    "address": "custodianAddr3",
    "balance": "10",
  },
]
`;

exports[`getLastOperationTime 1`] = `100000`;

exports[`getModerator 1`] = `"moderator"`;

exports[`getOperationCoolDown 1`] = `100000`;

exports[`getPoolAddresses 1`] = `
Array [
  Object {
    "address": "coldAddr1",
    "balance": "10",
  },
  Object {
    "address": "coldAddr2",
    "balance": "10",
  },
  Object {
    "address": "coldAddr3",
    "balance": "10",
  },
]
`;

exports[`getPoolAddresses 2`] = `
Array [
  Object {
    "address": "hotAddr1",
    "balance": "10",
  },
  Object {
    "address": "hotAddr2",
    "balance": "10",
  },
  Object {
    "address": "hotAddr3",
    "balance": "10",
  },
]
`;

exports[`getStates 1`] = `
Object {
  "custodianContractAddrs": Array [
    Object {
      "address": "custodianContractAddr1",
      "balance": 10,
    },
  ],
  "isStarted": true,
  "lastOperationTime": 123456789,
  "operationCoolDown": 1000,
  "otherContractAddrs": Array [
    Object {
      "address": "otherContractAddr1",
      "balance": 10,
    },
  ],
  "poolAddrsCold": Array [
    Object {
      "address": "hotAddr1",
      "balance": 10,
    },
  ],
  "poolAddrsHot": Array [
    Object {
      "address": "coldAddr1",
      "balance": 10,
    },
  ],
  "votingData": Object {
    "started": 123456789,
    "totalVoters": 100,
    "votedAgainst": 3,
    "votedFor": 4,
  },
  "votingStage": "Moderator",
}
`;

exports[`getVotingData 1`] = `
Object {
  "started": 123456789000,
  "totalVoters": "5",
  "votedAgainst": "5",
  "votedFor": "10",
}
`;

exports[`getVotingStage 1`] = `"Moderator"`;

exports[`isStarted 1`] = `true`;

exports[`removeAddress 1`] = `
Array [
  Array [
    "addr",
    1,
  ],
]
`;

exports[`removeAddress, readOnly 1`] = `"Read Only Mode"`;

exports[`startContractVoting, readOnly 1`] = `"Read Only Mode"`;

exports[`startContractVoting, with account 1`] = `
Array [
  Array [
    "candidateAdddr",
  ],
]
`;

exports[`startContractVoting, without account 1`] = `
Array [
  Array [
    "candidateAdddr",
  ],
]
`;

exports[`startManager, readOnly 1`] = `"Read Only Mode"`;

exports[`startManager, with account 1`] = `
Array [
  Array [],
]
`;

exports[`startManager, without account 1`] = `
Array [
  Array [],
]
`;

exports[`startModeratorVoting, readOnly 1`] = `"Read Only Mode"`;

exports[`startModeratorVoting, with account 1`] = `
Array [
  Array [],
]
`;

exports[`startModeratorVoting, without account 1`] = `
Array [
  Array [],
]
`;

exports[`terminateByTimeout, readOnly 1`] = `"Read Only Mode"`;

exports[`terminateByTimeout, with account 1`] = `
Array [
  Array [],
]
`;

exports[`terminateByTimeout, without account 1`] = `
Array [
  Array [],
]
`;

exports[`terminateContractVoting, with account 1`] = `
Array [
  Array [],
]
`;

exports[`terminateContractVoting, without account 1`] = `
Array [
  Array [],
]
`;

exports[`terminateContractVting, readOnly 1`] = `"Read Only Mode"`;

exports[`vote, readOnly 1`] = `"Read Only Mode"`;

exports[`vote, with account 1`] = `
Array [
  Array [
    true,
  ],
]
`;

exports[`vote, without account 1`] = `
Array [
  Array [
    true,
  ],
]
`;
